### DISCUSSION TITLE
Request stream only if a client is connected

### DISCUSSION THREAD

**Discussion by Barabba11** [bluenviron/mediamtx#553] [Questions]
Hi, I'm using it as a bridge, for the OBS RTSP server plugin ( https://github.com/iamscottxu/obs-rtspserver/ ) ,  it works ok. That plugin start the CPU encoding only when a client is connected, that saves power and life of CPU, unfortunately it doesn't force to connect with TCP only so I'm using your proxy, but your proxy always connect to it loading the CPU.

May you please implement to request from source the stream only when a client is connectiong to your proxy?
Thank you a lot!

**Comment by aler9** [comment]
It's enough to set the `sourceOnDemand` parameter to `yes`, as written in the configuration file:

```yml
paths:
  obs:
    source: rtsp://obs-rtspserver-address
    sourceOnDemand: yes
```

**Comment by Barabba11** [comment]
Thanks for answer, if I just remove # from default config file and instead of no I type yes I can't execute more the program

C:\Users\video2\Desktop\shttps\www>cd..

C:\Users\video2\Desktop\shttps>rtsp-simple-server
ERR: yaml: unmarshal errors:
  line 156: cannot unmarshal !!bool `yes` into conf.PathConf

C:\Users\video2\Desktop\shttps>

**Reply by aler9** [reply]
Just replace the configuration file with the configuration i posted above.

**Reply by Barabba11** [reply]
thank you, works as expected )
in my case I used:
```
paths:
  proxied:
    source: rtsp://127.0.0.1:501/live
    sourceOnDemand: yes
rtspAddress: :500
```

**Comment by Barabba11** [comment]
I've noticed errors, unfortunately :( 
I haven't adviced them before, probably because your program i able to retry to connect to server till it gets the stream. But now with the source on demand it's different, if it fails it reftects also on the player, which goes in timeout or your program will transmit an abort. I don't know why it happens this error 500, maybe is coming from OBS (or your program?) but there is a way to overcome it? For example retry faster or open the connecton in a more graceful way that obs doesn't refuse. Actually I'm open the connection with vlc direct to obs and it never fails, I don't know if there are aborts and different attempts of connection, I just know it works.
Can you please help me to solve this? Thank you
```
2021/09/07 20:44:38 I [0/0] rtsp-simple-server v0.17.2
2021/09/07 20:44:38 I [0/0] [path proxied] created
2021/09/07 20:44:38 I [0/0] [RTSP] UDP/RTP listener opened on :8000
2021/09/07 20:44:38 I [0/0] [RTSP] UDP/RTCP listener opened on :8001
2021/09/07 20:44:38 I [0/0] [RTSP] TCP listener opened on :500
2021/09/07 20:44:38 I [0/0] [RTMP] listener opened on :1935
2021/09/07 20:44:38 I [0/0] [HLS] listener opened on :8888
2021/09/07 20:49:06 I [0/0] [RTSP] [conn 192.168.24.1:37350] opened
2021/09/07 20:49:06 I [0/0] [path proxied] [rtsp source] started
2021/09/07 20:49:06 I [0/0] [path proxied] [rtsp source] ready
2021/09/07 20:49:06 I [0/0] [RTSP] [session 590851236] opened by 192.168.24.1:37
350
2021/09/07 20:49:06 I [0/1] [RTSP] [session 590851236] is reading from path 'pro
xied', 2 tracks with TCP
2021/09/07 20:49:09 I [0/1] [path proxied] [rtsp source] ERR: invalid status cod
e: 500 (Internal Server Error)
2021/09/07 20:49:09 I [0/0] [path proxied] [rtsp source] stopped
2021/09/07 20:49:09 I [0/0] [RTSP] [conn 192.168.24.1:37350] closed
2021/09/07 20:49:09 I [0/0] [RTSP] [session 590851236] closed
2021/09/07 20:49:22 I [0/0] [RTSP] [conn 192.168.24.1:37520] opened
2021/09/07 20:49:22 I [0/0] [path proxied] [rtsp source] started
2021/09/07 20:49:22 I [0/0] [path proxied] [rtsp source] ready
2021/09/07 20:49:22 I [0/0] [RTSP] [session 407918754] opened by 192.168.24.1:37
520
2021/09/07 20:49:22 I [0/1] [RTSP] [session 407918754] is reading from path 'pro
xied', 2 tracks with TCP
2021/09/07 20:49:25 I [0/1] [path proxied] [rtsp source] ERR: invalid status cod
e: 500 (Internal Server Error)
2021/09/07 20:49:25 I [0/0] [path proxied] [rtsp source] stopped
2021/09/07 20:49:25 I [0/0] [RTSP] [conn 192.168.24.1:37520] closed
2021/09/07 20:49:25 I [0/0] [RTSP] [session 407918754] closed

now with original config
2021/09/07 20:54:37 I [0/0] rtsp-simple-server v0.17.2
2021/09/07 20:54:37 I [0/0] [path runOnDemand] created
2021/09/07 20:54:37 I [0/0] [path runOnPublish] created
2021/09/07 20:54:37 I [0/0] [path runOnRead] created
2021/09/07 20:54:37 I [0/0] [path proxied] created
2021/09/07 20:54:37 I [0/0] [path proxied] [rtsp source] started
2021/09/07 20:54:37 I [0/0] [RTSP] TCP listener opened on :500
2021/09/07 20:54:37 I [0/0] [RTMP] listener opened on :1935
2021/09/07 20:54:37 I [0/0] [HLS] listener opened on :8888
2021/09/07 20:54:37 I [0/0] [path proxied] [rtsp source] ready
2021/09/07 20:54:40 I [0/0] [path proxied] [rtsp source] ERR: invalid status cod
e: 500 (Internal Server Error)
2021/09/07 20:54:45 I [0/0] [path proxied] [rtsp source] ready
2021/09/07 20:54:48 I [0/0] [path proxied] [rtsp source] ERR: invalid status cod
e: 500 (Internal Server Error)
2021/09/07 20:54:53 I [0/0] [path proxied] [rtsp source] ready
2021/09/07 20:55:43 I [0/0] [RTSP] [conn 192.168.24.1:40877] opened
2021/09/07 20:55:43 I [0/0] [RTSP] [session 191994447] opened by 192.168.24.1:40
877
2021/09/07 20:55:43 I [0/1] [RTSP] [session 191994447] is reading from path 'pro
xied', 2 tracks with TCP
2021/09/07 20:55:48 I [0/1] [RTSP] [conn 192.168.24.1:40877] closed
2021/09/07 20:55:48 I [0/0] [RTSP] [session 191994447] closed


```
from time to time it works, I need to play with VLC 2/3 times. WHen I stop and after I suddenly reconnect it works always, if I wait for some time, again the probem. The problem happens also with the original configuration file

**Comment by Barabba11** [comment]
Probably it is not related to your program, but you may help me try to overcome the issue.
Seams that RTSP server for the first seconds returns a 500, maybe it is caching/buffering, your program attempts too slowly trying to connect, it introduces another delay that makes the client's player to give out in timeout. 
Is it possible to speed up the attempts on reconnection? licke for example everly 0,2 seconds.. Thank you!

EDIT: I put also parameters from 10 to 1s but there is no improvement from this.
```
paths:
  proxied:
    source: rtsp://127.0.0.1:501/live
    sourceOnDemand: yes
    sourceOnDemandStartTimeout: 1s
    sourceOnDemandCloseAfter: 1s
rtspAddress: :500
```

[a comment by an automated account is not shown]
